src/HypreLinearSystem.C

```diff
@@ -140,83 +126,543 @@ HypreLinearSystem::beginLinearSystemConstruction()
+                                      & !(realm_.get_inactive_selector());
+  stk::mesh::BucketVector const& buckets = realm_.get_buckets(realm_.meta_data().side_rank(), s_owned);
+
+  if (numDof_==1) {
+    std::vector<HypreIntType> hids(0);
+    if (buckets.size()) {
+      const unsigned numNodes = (unsigned) (*buckets[0]).num_nodes(0);
+      hids.resize(numNodes);
+    }
+
+    for(size_t ib=0; ib<buckets.size(); ++ib) {
+      const stk::mesh::Bucket & b = *buckets[ib];
+      for ( stk::mesh::Bucket::size_type k = 0 ; k < b.size() ; ++k ) {
+
+	const unsigned numNodes = (unsigned) b.num_nodes(k);
+	
+	if (numNodes) {
+	  stk::mesh::Entity const * nodes = b.begin_nodes(k);
+	  
+	  /* save the hypre ids */
+	  for (unsigned i=0; i<numNodes; ++i) hids[i] = get_entity_hypre_id(nodes[i]);
+	  
+	  /* fill owned/shared 1 Dof */
+	  fill_owned_shared_data_structures_1DoF(numNodes, hids);
+	}
+      }
+    }
+  } else {
+    std::vector<HypreIntType> hids(0);
+    std::vector<HypreIntType> columns(0);
+    if (buckets.size()) {
+      const unsigned numNodes = (unsigned) (*buckets[0]).num_nodes(0);
+      hids.resize(numNodes);
+      columns.resize(numNodes*numDof_);
+    }
+    
+    for(size_t ib=0; ib<buckets.size(); ++ib) {
+      const stk::mesh::Bucket & b = *buckets[ib];
+      for ( stk::mesh::Bucket::size_type k = 0 ; k < b.size() ; ++k ) {
+	
+	const unsigned numNodes = (unsigned) b.num_nodes(k);
+	
+	if (numNodes) {
+	  stk::mesh::Entity const * nodes = b.begin_nodes(k);
+	  
+          /* save the hids and columns */
+	  fill_hids_columns(numNodes, nodes, hids, columns);
+
+	  /* fill owned/shared for more than 1 Dof */
+	  fill_owned_shared_data_structures(numNodes, hids, columns);
+	}
+      }
+    }
+  }
+
+#ifdef HYPRE_LINEAR_SYSTEM_TIMER
+  gettimeofday(&_stop, NULL);
+  double msec = (double)(_stop.tv_usec - _start.tv_usec) / 1.e3 + 1.e3*((double)(_stop.tv_sec - _start.tv_sec));
+  printf("rank_=%d EqnName=%s : %s %s %d : dt=%1.5lf\n",rank_,name_.c_str(),__FILE__,__FUNCTION__,__LINE__,msec);
+#endif
 }
 
+
 void
-HypreLinearSystem::buildEdgeToNodeGraph(
-  const stk::mesh::PartVector&)
+HypreLinearSystem::buildEdgeToNodeGraph(const stk::mesh::PartVector& parts)
 {
+#ifdef HYPRE_LINEAR_SYSTEM_TIMER
+  /* record the start time */
+  struct timeval _start, _stop;
+  gettimeofday(&_start, NULL);
+#endif
+
   beginLinearSystemConstruction();
+  stk::mesh::MetaData & metaData = realm_.meta_data();
+  const stk::mesh::Selector s_owned = metaData.locally_owned_part()
+                                      & stk::mesh::selectUnion(parts)
+                                      & !(realm_.get_inactive_selector());
+  stk::mesh::BucketVector const& buckets = realm_.get_buckets(stk::topology::EDGE_RANK, s_owned);
+
+  if (numDof_==1) {
+    std::vector<HypreIntType> hids(0);
+    if (buckets.size()) {
+      const unsigned numNodes = (unsigned) (*buckets[0]).num_nodes(0);
+      hids.resize(numNodes);
```

Comment:
This one should technically be after line 351 also. Edges are less likely to have different node counts than sides, but if we do high-order elements you could get some 3-node edges...

Reply:
fixed